yarl/_url.py

```diff
@@ -418,18 +417,31 @@ def __init_subclass__(cls):
 
     def __str__(self) -> str:
         val = self._val
+        scheme, netloc, path, query, fragment = val
         if not val.path and self.absolute and (val.query or val.fragment):
-            val = val._replace(path="/")
-        if (
-            explicit_port := self.explicit_port
-        ) is not None and explicit_port == self._default_port:
+            path = "/"
+        if (port := self.explicit_port) is not None and port == self._default_port:
             # port normalization - using None for default ports to remove from rendering
             # https://datatracker.ietf.org/doc/html/rfc3986.html#section-6.2.3
-            netloc = self._make_netloc(
-                self.raw_user, self.raw_password, self.host_subcomponent, None
-            )
-            val = val._replace(netloc=netloc)
-        return urlunsplit(val)
+            host = self.host_subcomponent
+            netloc = self._make_netloc(self.raw_user, self.raw_password, host, None)
+        return self._unsplit_result(scheme, netloc, path, query, fragment)
+
+    @staticmethod
+    def _unsplit_result(
+        scheme: str, netloc: str, url: str, query: str, fragment: str
+    ) -> str:
+        """Unsplit a URL without any normalization."""
+        if netloc or (scheme and scheme in USES_AUTHORITY) or url[:2] == "//":
+            if url and url[:1] != "/":
+                url = f"//{netloc or ''}/{url}"
+            else:
+                url = f"//{netloc or ''}{url}"
+        if scheme:
+            url = f"{scheme}:{url}"
+        if query:
+            url = f"{url}?{query}"
+        return f"{url}#{fragment}" if fragment else url
```

Comment:
Might be a lot faster to make a list and push and insert than join at the end.

Could be done in a future PR so we get a clean profile to prove the theory